section_synonym.go

```diff
@@ -0,0 +1,740 @@
+//  Copyright (c) 2024 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zap
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/RoaringBitmap/roaring/roaring64"
+	index "github.com/blevesearch/bleve_index_api"
+	seg "github.com/blevesearch/scorch_segment_api/v2"
+	"github.com/blevesearch/vellum"
+)
+
+func init() {
+	registerSegmentSection(SectionSynonymIndex, &synonymIndexSection{})
+	invertedIndexExclusionChecks = append(invertedIndexExclusionChecks, func(field index.Field) bool {
```

Comment:
Need to check if these exclusion checks are needed for the vector section as well - if yes/no, this seems quite a brittle way to support this going forward as we add more sections and we must come up with something better.

Reply:
I don't think an `exclusion` check is required for the vector/synonym sections ...
since they have `inclusion` checks each. The invertedTextIndex is a `catch-all` section that must process every field unless excluded (similar to default in switch-case). Moving forward, any new section will have to specify a special field type that only it can process.

The vector section can only process index.VectorField:

https://github.com/blevesearch/zapx/blob/3609ffc5f71b1d99026454ba60bf563f16cbe6ac/section_faiss_vector_index.go#L54-L57

The synonym section can only process index.SynonymField:
https://github.com/blevesearch/zapx/blob/3609ffc5f71b1d99026454ba60bf563f16cbe6ac/section_synonym.go#L387-L390

Reply:
Ok, then would you add a NOTE to where we defined these exclusion checks that why this is applicable _only_ to InvertedTextIndexSection.

Reply:
done